Add support for serializing top-level arrays to JSON

jeffwidman opened the pull request:
Major discussion of the issue is in #248. 

I'm tired of working around it, so just like to get it fixed. 

This solution is similar to #1402 & #1209 with a few changes:
* I found it unexpected that in #1402 to convert a python list to a top level array you had to pass `jsonify(*l)`. I changed it so that both `jsonify(1,2,3)` and `jsonify([1,2,3])` serialize to `[1,2,3]` as I think that's the more intuitive behavior.
* Per the discussion in #1209 & #248, I didn't create a separate config variable to modify this behavior. 

I know it'd be nice to incorporate a fix for #1443 in here as well, but I don't fully understand the issue there. So let's get this merged, and then tackle that issue in a followup PR.

Fix #170, #248, #510, #673, #1177 and check one of the boxes on #1182

ThiefMaster commented:
Shouldn't jsonify with both args and kwargs fail?

jeffwidman commented:
I don't know. Happy to change that if it should.

ThiefMaster commented on flask/json.py:
```diff
@@ -242,11 +242,21 @@ def get_current_user():
 
+    try:
+        data = dict(*args, **kwargs)
+    except:
```
A bare `except:` is almost never appropriate - here it isn't for sure!

jeffwidman replied:
Good call

ThiefMaster commented:
I wonder if the starargs-like syntax is actually useful.. If you have a fixed/low number of list entries, chances are good that you should be using a dict instead.

jeffwidman commented:
> If you have a fixed/low number of list entries, chances are good that you should be using a dict instead.

Why is that? 

I've got API endpoints like `/categories/` and `/users/` that I want to return lists of items. I don't understand why a dict would make more sense than an array even if I've only got 4-5 categories.

ThiefMaster commented:
No, I meant the cases where you would do `jsonify(1,2,3)` - if it's a list of users, you pass a list there, but you won't do `jsonify(u1, u2, u3)`

jeffwidman commented:
Agreed. 

`json.dumps()` only accepts the first item and appears to silently drop the others, so I considered copying that. It might be more natural as most people are used to the `dumps()` api. 

However, `jsonify()` currently accepts `**kwargs`, so I decided handling `*args` this way was more internally consistent.

Overall, I think the divergence in accepted input between `json.dumps()` and `jsonify()` creates more confusion than convenience. Right now I sometimes forget which one supports `kwargs` and which supports arrays. And I'm not the only one who has [trouble remembering](http://stackoverflow.com/questions/7907596/json-dumps-vs-flask-jsonify). 

Splitting behavior so `jsonify()` rejects `*args` but happily handles `**kwargs` just adds to the confusion, so we should pick either of the following:
1. Make `jsonify()` internally consistent so it accepts both `*args` and `**kwargs`. 
2. Make `jsonify()` consistent with `dumps()`, meaning they both only accept a single argument. This implies a long term plan to deprecate  `jsonify(**kwargs)` so that it throws an exception just like `dumps(**kwargs)`.

I don't care strongly, but if I had to pick I'd choose the latter because it's simpler to think of `jsonify()` as a `json.dumps()` that adds Flask niceties like the response class and serializing datetimes/UUIDs.

justanr commented:
:+1: on being consistent with json.dumps, especially since it's a widely copied API.

ThiefMaster commented:
BTW, I think the reason why `jsonify` uses `*args` instead of just a single `obj` argument is to allow an argument with that name name to be used as a kwarg.

mitsuhiko commented:
Maybe the solution to this mess is to introduce a new function that just invokes `json.dumps` and makes a response. Same behavior as `json.dumps` but with different name and makes a response object.

jeffwidman commented:
> Maybe the solution to this mess is to introduce a new function that just invokes json.dumps and makes a response. Same behavior as json.dumps but with different name and makes a response object.

Is there any reason to create a new function vs just modifying the existing `jsonify()` function? 

It breaks backwards compatibility, but for a 1.0 release that should be fine. Also, I know in my own projects, I've never intentionally used `jsonify(**kwargs)`, I've just always passed it a dict, and I suspect most people use it similarly.

It just seems the path of least surprise to stay with the current `jsonify()` function name and remove the `**kwargs` support.

ThiefMaster commented:
Breaking kwargs in `jsonify` would be a terrible idea. It's the kind of breakage that provides pretty much no real value and just annoys people (including myself ;)).

jeffwidman commented:
I hear you. :-)

My vote is to stick with my original implementation where both `jsonify(1,2,3)` and `jsonify([1,2,3])` serialize to `[1,2,3]`. That behavior is a bit more loose than `dumps()`, but I think it's more convenient for the user and I doubt it'd trip anyone up unexpectedly.

> Shouldn't jsonify with both args and kwargs fail?

This is still a valid question, but I don't want it to delay merging this PR. For now, let's just stick with Flask's current behavior which is to accept both `args` and `kwargs` and jsonify them to a dict. Changing that behavior can be discussed in a separate PR.

jeffwidman commented:
Any other feedback @mitsuhiko @davidism ? 

Would like to see this merged so I don't have to keep working around it. 

Supporting top-level arrays was already discussed ad nauseum in the issues I linked to in the OP, so this PR should be relatively non-controversial since it's just implementing behavior that already has been decided on.

ThiefMaster commented:
What about a test case for all the important cases on how to call `jsonify`? It'd be a shame to break any of them (kwargs / dict) without noticing  it at some point

jeffwidman commented:
Good idea. Let me see what I can do.

davidism commented on flask/json.py:
```diff
@@ -222,8 +222,8 @@ def get_current_user():
 
-    For security reasons only objects are supported toplevel.  For more
-    information about this, have a look at :ref:`json-security`.
+    Support for serializing top-level arrays was added in Flask 1.0. For
```
This should use a [`versionchanged`](http://sphinx-doc.org/markup/para.html#directive-versionadded) block.

davidism commented:
This needs a test case and docs to demonstrate exactly what is accepted by the function now.

Note that as a side effect, simple types are accepted as well (`True`, `3.14`, etc.), so there should be a test/doc for that too.  Test both a single and multi character string.

Not sure if I like that `*args` will be treated as an array, if it was me I'd make it always explicit, but it's not a huge deal.

jeffwidman commented:
Some updates:
1. Added some tests
2. Moved around the existing `jsonify()` tests so they're located next to each other--before they were a little scattered. Other than relocating, I didn't make any changes to existing tests.
3. Expanded the docs for `jsonify()`. I'm not very familiar with RST/sphinx, so there might be some formatting issues I missed.
4. I also restructured the actually code from a try/except block to a if clause. I thought the way it was previously structured left some confusion around when the `*args`/`**kwargs` behavior would throw an exception and switch from dict to array, so I made everything more explicit and linear.

Feedback welcome.

jeffwidman commented:
Reminder to take a look at this when you get a chance @davidism or @mitsuhiko

jeffwidman commented:
@davidism any chance you can review this weekend?

davidism commented:
I'll get to this on Sunday.

davidism commented:
OK, Monday morning. :wink: Thanks for your work on this.  I might clean up the wording a bit at some point, but other than that this looks good.

ThiefMaster commented on flask/json.py:
```diff
@@ -242,11 +256,21 @@ def get_current_user():
+        raise TypeError(
+            "jsonify() behavior undefined when passed both args and kwargs"
+        )
+    elif len(args) == 1: # single args are passed directly to dumps()
```
PEP8 issue, there should be two spaces before `#`

jeffwidman replied:
Oh, I didn't know that. I've been meaning to setup a pep8 checker on my local workflow, just haven't gotten to it.

jeffwidman replied:
Fixed in #1703

ThiefMaster commented on flask/json.py:
```diff
@@ -242,11 +256,21 @@ def get_current_user():
+    elif args: # convert multiple args into an array
+        data = list(args)
+    else: # convert kwargs to a dict
+        data = dict(kwargs)
```
isn't that a bit pointless? `kwargs` is always a dict

davidism replied:
And `args` is always a list.  Yeah, I'll clean that up.

jeffwidman replied:
Oh yeah. Good catch @ThiefMaster

ThiefMaster replied:
args is actually a tuple

jeffwidman replied:
Fixed in #1703

jeffwidman commented:
Thanks for merging @davidism, really nice to see this finally land. Also, did the best I could, but improvements to my wording are always appreciated.

jeffwidman commented on flask/json.py:
```diff
@@ -199,10 +199,22 @@ def htmlsafe_dump(obj, fp, **kwargs):
+    :func:`dumps`.
+    3. Multiple keyword arguments: Converted to a dict before being passed to
+    :func:`dumps`.
+    4. Both args and kwargs: Behavior undefined and will throw an exception.
```
I'm not sure why, but sphinx is not treating bullet points 2, 3, and 4 as list items. See the problem here: http://flask.pocoo.org/docs/dev/api/#flask.json.jsonify

wlhlm replied:
I think you have to line up the beginning of the lines for each point (eg Multiple and \:func\: have to be in the same vertical column.

http://docutils.sourceforge.net/docs/user/rst/quickref.html#enumerated-lists

jeffwidman replied:
Thanks, fixed in #1703

jeffwidman commented on flask/json.py:
```diff
@@ -222,8 +234,10 @@ def get_current_user():
+
+    .. versionchanged:: 1.0
+       Added support for serializing top-level arrays. This introduces a
+       security risk in ancient browsers. See :ref:`json_security` for details.
```
For some reason Sphinx isn't hyperlinking `json_security` on the generated docs. I'm not sure why. See the problem here: http://flask.pocoo.org/docs/dev/api/#flask.json.jsonify

ivan-leontiev replied:
Should be ``:ref:`json-security``` instead.

jeffwidman replied:
Many thanks! Just pushed a fix to #1703

ivan-leontiev commented on CHANGES:
```diff
@@ -8,6 +8,9 @@ Version 1.0
 
+- Added support to serializing top-level arrays to :func:`flask.jsonify`. This
+  introduces a security risk in ancient browsers. See
+  :ref:`json_security` for details.
```
The same here (json-security)

jeffwidman replied:
Good catch. I just added the fix to #1703.

visnup commented on docs/security.rst:
```diff
@@ -95,81 +95,12 @@ the form validation framework, which does not exist in Flask.
+vulnerabilities
+<https://github.com/mitsuhiko/flask/issues/248#issuecomment-59934857>`_, so
+this behavior was changed and :func:`~flask.jsonify` now supports serializing
+arrays.
```
:heart_eyes:
